### PR TITLE
raftstore: change the region merge flow path to fix some bugs

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: Liqi Geng <gengliqiii@gmail.com>

###  What have you changed?

Change the merge flow path to fix https://github.com/tikv/tikv/issues/6460 and the bug in https://github.com/tikv/tikv/issues/5981 comment.
1.   `exec_commit_merge` in target apply fsm and send `CatchUpLogs` to source peer fsm
2.   `on_catch_up_logs_for_merge` in source peer fsm
3.   if the source peer has already executed the corresponding `on_ready_prepare_merge`, set pending_remove and jump to step 6
4.   ... (raft append and apply logs)
5.   `on_ready_prepare_merge` in source peer fsm and set pending_remove (means source region has finished applying all logs)
6.   `logs_up_to_date_for_merge` in source apply fsm (destroy its apply fsm and send Noop to trigger the target apply fsm)
7.   resume `exec_commit_merge` in target apply fsm
8.   `on_ready_commit_merge` in target peer fsm and send `MergeResult` to source peer fsm
9.   `on_merge_result` in source peer fsm (destroy itself)

The first failpoint test can detect a lost-data bug in https://github.com/tikv/tikv/issues/6460.
The second one can detect a panic bug in https://github.com/tikv/tikv/issues/5981 comment.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No

###  Does this PR affect `tidb-ansible`?
No

###  Refer to a related PR or issue link (optional)
https://github.com/tikv/tikv/issues/6460
https://github.com/tikv/tikv/issues/5981

### CODE REVIEW

**Comment by gengliqi** [comment]
Note that the failpoint `before_send_rollback_merge_1003` parse data in `if` expression.
And I find that this `if` expression will run all the time when the failpoint feature is open.
I think the `if` expression shouldn't run if this failpoint is not open.
The result is all of the failpoint test may slow or even failed because of running `before_send_rollback_merge_1003`'s `if` expression.

File: components/test_raftstore/src/pd.rs [hunk 1]
```diff
@@ -868,23 +868,25 @@ impl TestPdClient {
...
 
-        for _ in 1..500 {
+        self.check_merged_timeout(from, 5);
+    }
+
+    pub fn check_merged(&self, from: u64) -> bool {
+        self.get_region_by_id(from).wait().unwrap().is_none()
+    }
+
+    pub fn check_merged_timeout(&self, from: u64, secs: u64) {
```
Comment (BusyJay) [inline components/test_raftstore/src/pd.rs]: Use `Duration` instead.

File: src/raftstore/store/fsm/apply.rs [hunk 2]
```diff
@@ -1768,14 +1766,16 @@ impl ApplyDelegate {
...
     // So, the merge process order would be:
-    // 1. `exec_commit_merge` in target apply worker
-    // 2. `catch_up_logs_for_merge` in source apply worker (check whether need to catch up logs)
-    // 3. `on_ready_catch_up_logs` in source raftstore
-    // 4. ... (raft append and apply logs)
-    // 5. `on_ready_prepare_merge` in source raftstore (means source region has finished applying all logs)
-    // 6. `catch_up_logs_for_merge` in source apply worker (destroy itself and send LogsUpToDate)
-    // 7. resume `exec_commit_merge` in target apply worker
-    // 8. `on_ready_commit_merge` in target raftstore
+    // 1.   `exec_commit_merge` in target applyfsm
```
Comment (BusyJay) [inline src/raftstore/store/fsm/apply.rs]: s/applyfsm/apply fsm/

File: src/raftstore/store/fsm/apply.rs [hunk 3]
```diff
@@ -2511,7 +2512,7 @@ impl ApplyFsm {
     fn resume_pending_merge(&mut self, ctx: &mut ApplyContext) -> bool {
         match self.delegate.wait_merge_state {
             Some(ref state) => {
-                let source_region_id = state.logs_up_to_date.load(Ordering::SeqCst);
+                let source_region_id = state.logs_up_to_date.load(Ordering::Acquire);
```
Comment (BusyJay) [inline src/raftstore/store/fsm/apply.rs]: Unnecessary change.

File: src/raftstore/store/fsm/apply.rs [hunk 4]
```diff
@@ -2593,18 +2566,20 @@ impl ApplyFsm {
...
             "source logs are all applied now";
             "region_id" => region_id,
             "peer_id" => self.delegate.id(),
         );
+        // The source peerfsm will be destroyed when the target peer executes `on_ready_commit_merge` 
+        // and sends `merge result` to the source peerfsm.
+        self.destroy(ctx);
+        catch_up_logs
+            .logs_up_to_date
+            .store(region_id, Ordering::Release);
```
Comment (BusyJay) [inline src/raftstore/store/fsm/apply.rs]: Use `SeqCst` instead.

**Comment by BusyJay** [comment]
> .. `cond` expression shouldn't run if its failpoint is not open.

Because access to failpoint holds a lock, executing condition check inside lock is a bad idea. Condition check is supposed to be cheap enough.

File: src/raftstore/store/fsm/apply.rs (hunk 4 above)
Comment (gengliqi) [reply src/raftstore/store/fsm/apply.rs]: The sequentially-consistent order is unnecessary. This is a typical scenario using acquire and release semantics.

Comment (BusyJay) [reply src/raftstore/store/fsm/apply.rs]: In TiKV, we prefer to use `SeqCst` in all places unless it's performance critical. The policy is to save mental analyze.

File: src/raftstore/store/peer.rs [hunk 5]
```diff
@@ -436,7 +437,6 @@ impl Peer {
         } else {
             initialized
         };
-        self.pending_remove = true;
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why remove the mark?

File: src/server/server.rs [hunk 6]
```diff
@@ -342,10 +342,15 @@ mod tests {
         }
     }
 
-    fn is_unreachable_to(msg: &SignificantMsg, region_id: u64, to_peer_id: u64) -> bool {
-        *msg == SignificantMsg::Unreachable {
+    fn is_unreachable_to(msg: &SignificantMsg, region_id_: u64, to_peer_id_: u64) -> bool {
+        if let SignificantMsg::Unreachable {
```
[the remark this answers is not in the document]
Comment (gengliqi) [reply src/server/server.rs]: I move `CatchUpLogs` to `SignificantMsg`. The atomic variable in `CatchUpLogs` can't be compared with each other.

File: src/raftstore/store/peer.rs (hunk 5 above)
Comment (gengliqi) [reply src/raftstore/store/peer.rs]: I move this mark to `handle_destroy_peer`. 
It's more clear to set this mark before really doing destroy things.
The `maybe_destroy` function should only check whether it is valid to destroy and change nothing.

File: src/raftstore/store/fsm/apply.rs [hunk 7]
```diff
@@ -1768,14 +1766,16 @@ impl ApplyDelegate {
...
-    // 8. `on_ready_commit_merge` in target raftstore
+    // 1.   `exec_commit_merge` in target apply fsm
+    // 2.   `on_catch_up_logs_for_merge` in source peer fsm
+    // 3.   if the source peer has already executed the corresponding `on_ready_prepare_merge`, set pending_remove and jump to step 6
+    // 4.   ... (raft append and apply logs)
+    // 5.   `on_ready_prepare_merge` in source peer fsm and set pending_remove (means source region has finished applying all logs)
+    // 6.   `logs_up_to_date_for_merge` in source apply fsm (destroy its apply fsm and send Noop to trigger the target apply fsm)
+    // 7.   resume `exec_commit_merge` in target apply fsm
+    // 8.   `on_ready_commit_merge` in target peer fsm
+    // 9.   `on_merge_result` in source peer fsm (destroy its peer fsm and send Noop to trigger the target peer fsm)
```
Comment (Connor1996) [inline src/raftstore/store/fsm/apply.rs]: already trigger `on_ready_commit_mege` in step 8, why send `Noop` again?

Comment (Connor1996) [reply src/raftstore/store/fsm/apply.rs]: seems the first time calling of `on_ready_commit_merge` is to send `MergeResult`, and do actual work for the second calling.

Comment (Connor1996) [reply src/raftstore/store/fsm/apply.rs]: seems the first time calling of `on_ready_commit_merge` is to send `MergeResult`, and do actual work for the second calling?

File: src/raftstore/store/fsm/peer.rs [hunk 8]
```diff
@@ -2046,21 +1923,22 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
-            PeerMsg::CasualMessage(CasualMessage::MergeResult {
+            PeerMsg::SignificantMsg(SignificantMsg::MergeResult {
                 target: self.fsm.peer.peer.clone(),
                 stale: false,
             }),
         ) {
-            info!(
+            // TODO: need to remove "are we shutting down", it should panic
+            // if we are not in shut-down state
+            warn!(
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: Let's keep it as info until the TODO is resolved.

File: src/raftstore/store/peer.rs [hunk 9]
```diff
@@ -1169,7 +1155,16 @@ impl Peer {
             "region_id" => self.region_id,
             "peer_id" => self.peer.get_id(),
         );
-
+        {
+            let before_handle_raft_ready_1003 = || {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why put it into a closure?

File: tests/failpoints/cases/test_merge.rs [hunk 10]
```diff
@@ -653,77 +637,220 @@ fn test_node_merge_restart_after_apply_premerge_before_apply_compact_log() {
...
     fail::remove(schedule_merge_fp);
 
-    // propose to left region and wait for merge to succeed conveniently
+    pd_client.check_merged_timeout(left.get_id(), Duration::from_secs(5));
+
     cluster.must_put(b"k123", b"v2");
     must_get_equal(&cluster.get_engine(3), b"k123", b"v2");
 }
+
+// Test if the data is complete when the source region executes the previous
```
Comment (BusyJay) [inline tests/failpoints/cases/test_merge.rs]: How about
```
/// Tests whether stale merge is rollback properly if it merge to the same target region again later.
```

File: tests/failpoints/cases/test_merge.rs [hunk 11]
```diff
@@ -653,77 +637,220 @@ fn test_node_merge_restart_after_apply_premerge_before_apply_compact_log() {
...
+    // Make peer 1003 can't receive PrepareMerge and RollbackMerge log
+    cluster.add_send_filter(IsolationFilterFactory::new(3));
+
+    cluster.try_merge(left.get_id(), right.get_id());
+
+    // Change right region's epoch to make this merge failed
+    cluster.must_split(&right, b"k8");
+    fail::remove(schedule_merge_fp);
+    // Wait for left region to rollback merge
+    sleep_ms(200);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_merge.rs]: How about put a key value instead?

File: tests/failpoints/cases/test_merge.rs [hunk 12]
```diff
@@ -653,77 +637,220 @@ fn test_node_merge_restart_after_apply_premerge_before_apply_compact_log() {
...
+    // Wait `after_send_to_apply_1003` timeout
+    sleep_ms(300);
+    fail::remove(after_send_to_apply_1003_fp);
+    // Check the data integrity
+    for i in 0..5 {
+        must_get_equal(&cluster.get_engine(3), format!("k2{}", i).as_bytes(), b"v3");
+    }
+}
+
+// Test if the source region is destroyed correctly in merging when meeting transfer leader.
```
Comment (BusyJay) [inline tests/failpoints/cases/test_merge.rs]: How about
```
/// Tests whether the source peer is destroyed correctly when transferring leader during committing merge.
///
/// In the previous merge flow, target peer deletes meta of source peer without marking it as pending remove.
/// If source peer becomes leader at the same time, it will panic due to corrupted meta.
```

File: tests/failpoints/cases/test_merge.rs [hunk 13]
```diff
@@ -653,77 +637,220 @@ fn test_node_merge_restart_after_apply_premerge_before_apply_compact_log() {
...
+
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_put(b"k3", b"v3");
+
+    let left = pd_client.get_region(b"k1").unwrap();
+    let right = pd_client.get_region(b"k2").unwrap();
+
+    let left_peer_1 = find_peer(&left, 1).unwrap().to_owned();
+    cluster.must_transfer_leader(left.get_id(), left_peer_1.clone());
+    sleep_ms(100);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_merge.rs]: Why sleep?

**Comment by youjiali1995** [comment]
Does it need to be merged into v3.1.0-beta.2? @gengliqi
